### PR TITLE
Added KP requirements for reasoning

### PR DESCRIPTION
**Pull request by cbizon**
Based on the discussion here: https://github.com/NCATSTranslator/TranslatorArchitecture/discussions/25

It's not clear to me that this is the best place for this- I think if things were further along, this might actually be part of the specification of either TRAPI or one of its operations, but for now this is a reasonable place to get agreement.  We can rearrange going forward.

If anybody wants to improve the language for clarity, please do so.

### CODE REVIEW

**Review by edeutsch — COMMENTED** [review]
I do not specifically or strongly object to this. But this seems like a big ask to me. And one that I find unexpected. I am intensely curious what the KPs think. It's pretty quiet out there. I expected wails of protest!

**Comment by cbizon** [comment]
It may be quiet because we've had much of this discussion here: https://github.com/NCATSTranslator/TranslatorArchitecture/discussions/25

Or it may be that people are unaware of this PR :)   On the last call, I promised to try to tag these issues better, so here we go:

@CaseyTa
@yakaboskic
@stevencox
@marcdubybroad
@vdancik
@rtroper
@bill-baumgartner
@newgene
@patrickkwang
@suihuang-ISB
@Rosinaweber
@andrewsu
@edeutsch
@MarkDWilliams
@webyrd
@balhoff

**Review by marcdubybroad — COMMENTED** [review]
Just saw the 'KPs must not call other KPs' part; realize it is not part of this pull request, but thought that if a KP can pull another KPs data and augment it or use it to augment their own data (computation, etc), whether that would be allowed.

**Review by marcdubybroad — APPROVED** [review]
The only reservation is the 'inverse predicate' segment. Having KPs /predicates call (or whatever replaces it) be as explicit as possible will make human free querying a lot simpler. So if a KP has A->treats->B, then they should specify B->is_treated_by->A if they allow that query. The more we keep things implicit them more we run the risk of having different interpretations/implementations and make querying harder to be automated.
Unclear if there is a use case where the above doesn't apply.

**Comment by cbizon** [comment]
@marcdubybroad In my mind, this PR is tantamount to saying that if a KP responds to A-treats->B, then it should also respond to an equivalent B-is_treated_by->A query.   Which of those two or both is advertised is I think an independent question, and is the point of this discussion: https://github.com/NCATSTranslator/TranslatorArchitecture/discussions/31

**Comment by sstemann** [comment]
@NCATSTranslator/architecture-core Arch Committee reps, please review/approve/disapprove/comment as reviewers on this PR.

File: README.md [hunk 1]
```diff
@@ -47,6 +47,11 @@ This README documents the current strawman architecture.  Changes must be made v
    1. The SRI will provide mapping tools to perform this conversion.
 1. ARAs and KPs may both score answers (provide scores in the message); ARAs are required to score answers.
 1. KPs should not call other KPs.
+1. KPs that implement the Translator Reasoner API must perform the following kinds of reasoning in answering queries:
+   1. Making identifiers more specific, e.g. responding to a query involving an entity with information related to a subclass of that entity.
+   1. Making categories in a query more specific. e.g. responding to a query for a Named Thing with a particular ChemicalSubstance.
```
Comment (patrickkwang) [inline README.md:52]:
```suggestion
   1. Making categories in a query more specific. e.g. responding to a query for a biolink:NamedThing with a particular biolink:ChemicalSubstance.
```

**Review by patrickkwang — APPROVED** [review]

**Review by bill-baumgartner — APPROVED** [review]

**Review by stevencox — APPROVED** [review]

**Review by balhoff — APPROVED** [review]

**Comment by rtroper** [comment]
We make use of the service provider group's tools for deploying our KGs as KPs. Any of these four requirements that could be handled at the KG level (e.g. the TSV files defining nodes and edges), we'd be able to meet ourselves. The others would have to be handled by the service provider group. Looking over these requirements, the only one I think we could feasibly do at the KG level is requirement 4. For each edge, we would just explicitly include an additional edge for the inverse in the knowledge graph itself (rather than relying on 'reasoning' logic at the KP level). So, I guess if @newgene were on board with this pull request, we would be on board.

**Comment by newgene** [comment]
@rtroper you are right. As detailed in @kevinxin90's comment here, https://github.com/NCATSTranslator/TranslatorArchitecture/discussions/25#discussioncomment-387337, Service Provider will handle reasoning at its TRAPI wrapper level, ideally through a centralized normalizing service. That way, the underlying KG src does not need to do anything special. The most important for data src is to provide as specific as possible annotations for their nodes and predicates.

**Review by newgene — APPROVED** [review]

**Review by andrewsu — APPROVED** [review]

**Review by rtroper — APPROVED** [review]

**Comment by rtroper** [comment]
> @rtroper you are right. As detailed in @kevinxin90's comment here, [#25 (reply in thread)](https://github.com/NCATSTranslator/TranslatorArchitecture/discussions/25#discussioncomment-387337), Service Provider will handle reasoning at its TRAPI wrapper level, ideally through a centralized normalizing service. That way, the underlying KG src does not need to do anything special. The most important for data src is to provide as specific as possible annotations for their nodes and predicates.

@newgene Thanks for the reply!

**Review by suihuang-ISB — APPROVED** [review]

**Review by edeutsch — APPROVED** [review]

**Review by edeutsch — COMMENTED** [review]
So the principles seem clear here and I approved, but implementation details need a little more deciding and definition I think. Here's one:

After the KP does some reasoning to answer via an inverted predicate, should/must the response be the native information or the inverted reasoned information as requested? for example, if ARA asks KP: does A treat B, and if KP locally stores B is_treated_by A, must the KP respond A treats B, or is the KP free to respond B is_treated_by A. i.e. is the ARA guaranteed to receive the answer in the same way that it asked the question? My reading of the current text seems to require that the KP perform reasoning to find information correctly, but does not clearly stipulate how the response must be phrased.

**Review by CaseyTa — APPROVED** [review]

**Review by yakaboskic — CHANGES_REQUESTED** [review]
First, thank you for tagging reps on issues, this is really helpful and I might not have seen this otherwise!

While I agree that making all identifiers, categories, predicates more specific is a good thing, I think the 3rd and 4th requirements are a little restrictive. As a KP that attempts to handle query context, I think allowing edges that are not directly being responded to but are providing context for the query calculation should still be allowed. Similarly, in an query graph that has more than one edge, what does an inversion mean? It seems this might be a push to try to limit KPs to one-hop functionality and I think that is extremely limiting/worrying.

**Comment by cbizon** [comment]
@edeutsch Agreed, this PR is written at a pretty high level and doesn't address your question.   I think we should agree on the principle and then work on the implementation.  Do you think that the best place to have that conversation is on the arch call, or TRAPI?

**Comment by cbizon** [comment]
@yakaboskic I don't think that this PR excludes either returning extra information (nodes/edges) or limits KPs to providing only one-hops.  That is, this PR does not attempt to describe the totality of what KPs do, but to instead provide some basics around how they must act.  So for instance, if an inverse of treats to treated_by occurs in the middle of a larger trapi query, then the KP should return the relevant information rather than punting.

Is there some wording that would make that clearer in the PR?  Or is there a deeper issue that I am missing?

**Comment by edeutsch** [comment]
> @edeutsch Agreed, this PR is written at a pretty high level and doesn't address your question. I think we should agree on the principle and then work on the implementation. Do you think that the best place to have that conversation is on the arch call, or TRAPI?

It could go in either, but I would suggest architecture since this group is already addressing this broader topic and is primed to consider finer details.

**Pull request merged**